src/proc/interpolator.rs

```diff
@@ -0,0 +1,119 @@
+    ///
+    /// All the shader languages Naga supports have similar rules:
+    /// perspective-correct, center-sampled interpolation is the default for any
+    /// binding that can vary, and everything else either defaults to flat, or
+    /// requires an explicit flat qualifier/attribute/what-have-you.
+    ///
+    /// [`Binding`]: super::Binding
+    pub fn apply_common_default_interpolation(&mut self) {
+        use crate::{Binding, ScalarKind, Type, TypeInner};
+
+        /// Choose a default interpolation for a function argument or result.
+        ///
+        /// `binding` refers to the `Binding` whose type is `ty`. If `ty` is a struct, then it's the
+        /// bindings of the struct's members that we care about, and the binding of the struct
+        /// itself is meaningless, so `binding` should be `None`.
+        fn default_binding_or_struct(binding: &mut Option<Binding>,
+                                     ty: Handle<Type>,
+                                     types: &mut Arena<Type>)
+        {
+            match types.get_mut(ty).inner {
+                // A struct. It's the individual members we care about, so recurse.
+                TypeInner::Struct { members: ref mut m, .. } => {
+                    // To choose the right interpolations for `members`, we must consult other
+                    // elements of `types`. But both `members` and the types it refers to are stored
+                    // in `types`, and Rust won't let us mutate one element of the `Arena`'s `Vec`
+                    // while reading others.
+                    //
+                    // So, temporarily swap the member list out its type, assign appropriate
+                    // interpolations to its members, and then swap the list back in.
+                    use std::mem::replace;
+                    let mut members = replace(m, vec![]);
+
+                    for member in &mut members {
+                        default_binding_or_struct(&mut member.binding, member.ty, types);
+                    }
+
+                    // Swap the member list back in. It's essential that we call `types.get_mut`
+                    // afresh here, rather than just using `m`: it's only because `m` was dead that
+                    // we were able to pass `types` to the recursive call.
+                    match types.get_mut(ty).inner {
+                        TypeInner::Struct { members: ref mut m, .. } => replace(m, members),
+                        _ => unreachable!("ty must be a struct"),
+                    };
+                }
+
+                // Some interpolatable type.
+                //
+                // GLSL has 64-bit floats, but it won't interpolate them. WGSL and MSL only have
+                // 32-bit floats. SPIR-V has 16- and 64-bit float capabilities, but Vulkan is vague
+                // about what can and cannot be interpolated.
+                TypeInner::Scalar { kind: ScalarKind::Float, width: 4 } |
```

Comment:
why are we checking for `width: 4` specifically in these cases? I imagine, once we have other widths, we'd need to change it anyway, and right now this processor is not the right place to validate the width, anyway.

Reply:
that's an interesting discovery! I wonder what other backends think of this.